src/v/security/gssapi_principal_mapper.cc

```diff
@@ -0,0 +1,474 @@
+
+#include <boost/algorithm/string/case_conv.hpp>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <charconv>
+
+namespace security {
+static constexpr const char* const gssapi_name_pattern
+  = R"(([^/@]*)(/([^/@]*))?@([^/@]*))";
+static constexpr const char* const non_simple_pattern = R"([/@])";
+static constexpr const char* const parameter_pattern = R"(([^$]*)(\$(\d*))?)";
+static constexpr const char* const rule_pattern
+  = R"(((DEFAULT)|((RULE:\[(\d*):([^\]]*)](\(([^)]*)\))?(s\/([^\/]*)\/([^\/]*)\/(g)?)?\/?(L|U)?))))";
+
+namespace detail {
+std::vector<gssapi_rule> parse_rules(
+  std::string_view default_realm,
+  std::optional<std::vector<ss::sstring>> unparsed_rules) {
+    static const RE2 rule_parser(rule_pattern, RE2::Quiet);
+
+    vassert(
+      rule_parser.ok(),
+      "Failed to build rule pattern regex: {}",
+      rule_parser.error());
+
+    if (!unparsed_rules || unparsed_rules.value().empty()) {
+        return {gssapi_rule(default_realm)};
+    }
+
+    std::vector<gssapi_rule> rv;
+    re2::StringPiece default_;
+    re2::StringPiece num_components_str;
+    re2::StringPiece format;
+    re2::StringPiece match_regex;
+    re2::StringPiece from_pattern;
+    re2::StringPiece to_pattern;
+    re2::StringPiece repeat;
+    re2::StringPiece upper_lower;
+    for (const auto& rule : *unparsed_rules) {
+        const re2::StringPiece rule_piece(rule.data(), rule.size());
+        if (!RE2::FullMatch(
+              rule_piece,
+              rule_parser,
+              nullptr,
+              &default_,
+              nullptr,
+              nullptr,
+              &num_components_str,
+              &format,
+              nullptr,
+              &match_regex,
+              nullptr,
+              &from_pattern,
+              &to_pattern,
+              &repeat,
+              &upper_lower)) {
+            throw std::runtime_error("GSSAPI: Invalid rule: " + rule);
+        }
+        if (!default_.empty()) {
+            rv.emplace_back(default_realm);
+        } else {
+            int num_components = std::numeric_limits<int>::max();
+            auto conv_rc = std::from_chars(
+              num_components_str.begin(),
+              num_components_str.end(),
+              num_components);
+            if (conv_rc.ec != std::errc()) {
+                throw std::runtime_error(
+                  "Invalid rule - Invalid value for number of components: "
+                  + num_components_str.as_string());
+            }
+            rv.emplace_back(
+              default_realm,
+              num_components,
+              format,
+              match_regex,
+              from_pattern,
+              to_pattern,
+              gssapi_rule::repeat{repeat == "g"},
+              gssapi_rule::make_lower{upper_lower == "L"},
+              gssapi_rule::make_upper{upper_lower == "U"});
+        }
+    }
+
+    return rv;
+}
+} // namespace detail
+
+gssapi_name::gssapi_name(
+  std::string_view service_name,
```

Comment:
nit: This might better be called `primary` as I imagine it could be a user or service.

Reply:
I was matching other source that was calling this "service_name" but I can replace it with primary if you prefer.